### ISSUE TITLE
Advice.DynamicValue.ForFixedValue does not support Class values

### ISSUE THREAD

**Issue by felixbarny** [raphw/byte-buddy#221]
With the previous implementation of Advice.DynamicValue, I was able to return a `Class` object in the `resolve` method: https://github.com/stagemonitor/stagemonitor/blob/7d0d8f625aba65e392df834329ee344be77e4e2d/stagemonitor-core/src/main/java/org/stagemonitor/core/metrics/annotations/ExceptionMeteredTransformer.java#L79

When trying to adapt to `Advice.DynamicValue.ForFixedValue`, I get this stack trace:

```
java.lang.IllegalStateException: Not a constant value: class java.lang.Exception
	at net.bytebuddy.asm.Advice$DynamicValue$ForFixedValue.resolve(Advice.java:8782)
	at net.bytebuddy.asm.Advice$Dispatcher$OffsetMapping$ForUserValue.resolve(Advice.java:4315)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved$ForMethodExit.resolve(Advice.java:7310)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved$ForMethodExit$WithExceptionHandler.resolve(Advice.java:7342)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved.bind(Advice.java:6795)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved$ForMethodExit$WithExceptionHandler.bind(Advice.java:7342)
	at net.bytebuddy.asm.Advice$AdviceVisitor.<init>(Advice.java:7514)
	at net.bytebuddy.asm.Advice$AdviceVisitor$WithExitAdvice.<init>(Advice.java:7761)
	at net.bytebuddy.asm.Advice$AdviceVisitor$WithExitAdvice$WithExceptionHandling.<init>(Advice.java:7979)
	at net.bytebuddy.asm.Advice.doWrap(Advice.java:524)
	at net.bytebuddy.asm.Advice.wrap(Advice.java:467)
	at net.bytebuddy.asm.AsmVisitorWrapper$ForDeclaredMethods$Entry.wrap(AsmVisitorWrapper.java:563)
	at net.bytebuddy.asm.AsmVisitorWrapper$ForDeclaredMethods$DispatchingVisitor.visitMethod(AsmVisitorWrapper.java:668)
	at net.bytebuddy.jar.asm.ClassVisitor.visitMethod(Unknown Source)
	at net.bytebuddy.asm.AsmVisitorWrapper$ForDeclaredMethods$DispatchingVisitor.visitMethod(AsmVisitorWrapper.java:664)
	at net.bytebuddy.jar.asm.ClassVisitor.visitMethod(Unknown Source)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining$RedefinitionClassVisitor.visitMethod(TypeWriter.java:3483)
	at net.bytebuddy.jar.asm.ClassReader.b(Unknown Source)
	at net.bytebuddy.jar.asm.ClassReader.accept(Unknown Source)
	at net.bytebuddy.jar.asm.ClassReader.accept(Unknown Source)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:3095)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:1693)
	at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:171)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Transformation$Simple$Resolution.apply(AgentBuilder.java:7396)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:7859)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:7824)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1100(AgentBuilder.java:7611)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:8249)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:8199)
	at java.security.AccessController.doPrivileged(Native Method)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:7750)
	at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:455)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:367)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.stagemonitor.alerting.annotation.SlaCheckCreatingClassPathScannerTest.init(SlaCheckCreatingClassPathScannerTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by raphw** [comment]
Ah, shoot. I moved the wrapping to the DSL. You can fix this by wrapping the result in a `TypeDescription.ForLoadedType`. Sorry for the regression!

**Comment by raphw** [comment]
As a background, I did a major rewrite of the `Advice` component. It does now use standard components what makes it much easier to write to values as the component now uses standard assignment like boxing etc. The generated byte code is also much more efficient. But this one slipped under.

The good news: With the rewrite, I think the `Advice` component has finally stabalized, I do not think there will be further deep changes.

**Comment by felixbarny** [comment]
Thx for the ~quick~ instant answer!

Unfortunately, I get a different error now:

```
		@Override
		protected Object doResolve(TypeDescription instrumentedType,
								   MethodDescription instrumentedMethod,
								   ParameterDescription.InDefinedShape target,
								   AnnotationDescription.Loadable<MeterExceptionsFor> annotation,
								   Assigner assigner, boolean initialized) {
			return new TypeDescription.ForLoadedType(instrumentedMethod.getDeclaredAnnotations().ofType(ExceptionMetered.class).loadSilent().cause());
		}
```

```
java.lang.IllegalStateException: An illegal stack manipulation must not be applied
	at net.bytebuddy.implementation.bytecode.StackManipulation$Illegal.apply(StackManipulation.java:47)
	at net.bytebuddy.implementation.bytecode.StackManipulation$Compound.apply(StackManipulation.java:214)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved$AdviceMethodWriter.doApply(Advice.java:6938)
	at net.bytebuddy.asm.Advice$Dispatcher$Delegating$Resolved$AdviceMethodWriter$ForMethodExit.apply(Advice.java:7106)
	at net.bytebuddy.asm.Advice$AdviceVisitor$WithExitAdvice.onUserEnd(Advice.java:7838)
	at net.bytebuddy.asm.Advice$AdviceVisitor.visitMaxs(Advice.java:7579)
	at net.bytebuddy.jar.asm.MethodVisitor.visitMaxs(Unknown Source)
	at net.bytebuddy.asm.Advice$AdviceVisitor.visitMaxs(Advice.java:7580)
	at net.bytebuddy.jar.asm.ClassReader.a(Unknown Source)
	at net.bytebuddy.jar.asm.ClassReader.b(Unknown Source)
	at net.bytebuddy.jar.asm.ClassReader.accept(Unknown Source)
	at net.bytebuddy.jar.asm.ClassReader.accept(Unknown Source)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:3095)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:1693)
	at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:171)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Transformation$Simple$Resolution.apply(AgentBuilder.java:7396)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:7859)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:7824)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1100(AgentBuilder.java:7611)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:8249)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:8199)
	at java.security.AccessController.doPrivileged(Native Method)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:7750)
	at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:455)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:367)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.stagemonitor.alerting.annotation.SlaCheckCreatingClassPathScannerTest.init(SlaCheckCreatingClassPathScannerTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by raphw** [comment]
Ah, this is indeed a bug, sorry for this. I will fix this ASAP.

**Comment by raphw** [comment]
Fixed in 1.5.7.

**Comment by felixbarny** [comment]
Thx, working nicely!
